mitxgraders/comparers/comparers.py

```diff
@@ -347,3 +347,81 @@ def vector_phase_comparer(comparer_params_evals, student_eval, utils):
     same_magnitude = utils.within_tolerance(expected_mag, student_mag)
 
     return in_span and same_magnitude
+
+class CorrellatedComparer(object):
+    """
+    CorrelatedComparer instances are callable objects used as comparer functions
+    in FormulaGrader problems. Unlike standard comparer functions, CorrellatedComparer
+    instances are given access to all parameter evaluations at once.
+
+    For example, a comparer function that decides whether the student input is a
+    nonzero constant multiple of the expected input would need to be a correlated
+    comparer so that it can determine if there is a linear relationship between
+    the student and expected samples.
+    """
+
+    def __init__(self, func):
+        self._comparer = func
+
+    def __call__(self, comparer_params_evals, student_evals, utils):
+        return self._comparer(comparer_params_evals, student_evals, utils)
+
+def make_constant_multiple_comparer(grade_decimal=0.5, msg='The submitted answer differs from the expected answer by a constant multiple'):
+    """
+    Makes a comparer function that tests whether student input is a constant multiple
+    of expected input, and if so, gives partial credit and displays a message.
+
+    Usage
+    =====
+
+    >>> from mitxgraders import FormulaGrader
+    >>> grader = FormulaGrader(
+    ...     answers={
+    ...         'comparer_params': ['m*c^2'],
+    ...         'comparer': make_constant_multiple_comparer(grade_decimal=0.75)
+    ...     },
+    ...     variables=['m', 'c']
+    ... )
+    >>> result = grader(None, '2*m*c^2')
+    >>> result == {
+    ...     'ok': 'partial',
+    ...     'msg': 'The submitted answer differs from the expected answer by a constant multiple',
+    ...     'grade_decimal': 0.75
+    ... }
+    True
+
+
+    Gives full credit / no credit appropriately:
+    >>> grader(None, 'm*c^3')['ok']
+    False
+    >>> grader(None, 'm*c^2')['ok']
+    True
+
+    Zero input is always marked wrong:
+    >>> grader(None, '0')['ok']
+    False
+    """
+
+
+    @CorrellatedComparer
+    def _comparer(comparer_params_evals, student_evals, utils):
+        student_eval_norm = np.linalg.norm(student_evals)/len(student_evals)
+        if is_nearly_zero(student_eval_norm, utils.tolerance, reference=student_evals):
+            return False
+
+        A = np.vstack([student_evals]).T
+        y = [c[0] for c in comparer_params_evals]
+        sol, residuals, _, _ = np.linalg.lstsq(A, y, rcond=-1)
+        coeff = sol[0]
+        error = np.sqrt(residuals)
+
+        if is_nearly_zero(error, utils.tolerance, reference=student_eval_norm):
+            if is_nearly_zero(coeff - 1, utils.tolerance, reference=student_eval_norm):
+                return True
+            return { 'grade_decimal': grade_decimal, 'msg': msg }
```

Comment:
Hmm. I note that this won't play nice with `num_failures`. I think this is ok, but it will need to be documented.